panopticapi/evaluation.py

```diff
@@ -83,9 +83,17 @@ def pq_compute_single_core(proc_id, annotation_set, gt_folder, pred_folder, cate
             print('Core: {}, {} from {} images processed'.format(proc_id, idx, len(annotation_set)))
         idx += 1
 
-        pan_gt = np.array(Image.open(os.path.join(gt_folder, gt_ann['file_name'])), dtype=np.uint32)
+        pan_gt_path = gt_ann['file_name']
+        if not os.path.split(pan_gt_path)[0]:
```

Comment:
It'd be great if we could minimise how far we diverge our fork from the original library - ideally, we'd raise a PR against the original project for any bug fixes and stick to the official version. 

Here, the previous behaviour always calls os.path.join() on the folder and file_name whereas now it's only doing so if file_name doesn't contain a "/" path delimiter... doesn't this restrict which paths can be specified rather than expand it?

os.path.join() will discard the previous path parts if you join on an absolute path:
```
>>> os.path.join("/derp/foo", "blah/whatever")
'/derp/foo/blah/whatever'
>>> os.path.join("/derp/foo", "/blah/whatever")
'/blah/whatever'
```

So shouldn't we just be specifying an absolute path file_name?

Reply:
> It'd be great if we could minimise how far we diverge our fork from the original library - ideally, we'd raise a PR against the original project for any bug fixes and stick to the official version.

Raising a PR is not going to do much. The original repo has not been updated in 2 years so my hopes aren't high.

> Here, the previous behaviour always calls os.path.join() on the folder and file_name whereas now it's only doing so if file_name doesn't contain a "/" path delimiter... doesn't this restrict which paths can be specified rather than expand it?

I don't think so. Absolute paths should still work just fine. Only file names shoud work just fine. Except now relative paths should also work. How is that a restriction?

> So shouldn't we just be specifying an absolute path file_name?

I'd really prefer using relative paths in the dataset_json (where gt_ann[file_name] comes from) so that the file doesn't have to be updated every time the dataset is moved to another location or new users download the data. It would be much neater to specify paths relative to the json file. 

Since panopticapi relies on json files to find relevant files, and the json file in use is the same dataset.json that is used for other things, I'm hesitant to manipulate the file just so we have one less edit on a poorly maintained package.

Reply:
> > It'd be great if we could minimise how far we diverge our fork from the original library - ideally, we'd raise a PR against the original project for any bug fixes and stick to the official version.
> 
> Raising a PR is not going to do much. The original repo has not been updated in 2 years so my hopes aren't high.
> 

Yeah, it might not be possible in this case and that's fine. I think we just need to be very reluctant, in general, to making lots of modifications to third-party libraries and having to maintain our own forks in the long-term. There's an increasing maintenance burden we're taking on here that we should avoid if at all possible.

> > Here, the previous behaviour always calls os.path.join() on the folder and file_name whereas now it's only doing so if file_name doesn't contain a "/" path delimiter... doesn't this restrict which paths can be specified rather than expand it?
> 
> I don't think so. Absolute paths should still work just fine. Only file names shoud work just fine. Except now relative paths should also work. How is that a restriction?
> 
> > So shouldn't we just be specifying an absolute path file_name?
> 
> I'd really prefer using relative paths in the dataset_json (where gt_ann[file_name] comes from) so that the file doesn't have to be updated every time the dataset is moved to another location or new users download the data. It would be much neater to specify paths relative to the json file.
> 
> Since panopticapi relies on json files to find relevant files, and the json file in use is the same dataset.json that is used for other things, I'm hesitant to manipulate the file just so we have one less edit on a poorly maintained package.

Ah, I was missing context that this is read directly from a config file rather than passed as a config object into the library. I guess it just seems strange that there's a `gt_folder` parameter which in the original code allows you to specify a parent folder under which you can nest relative paths from `gt_ann['file_name']` whereas now the relative paths will be from the working directory. If there's no avoiding it for how we're using the data then sounds good though - we definitely don't want to have to edit config whenever we move folders around.

Reply:
> Yeah, it might not be possible in this case and that's fine. I think we just need to be very reluctant, in general, to making lots of modifications to third-party libraries and having to maintain our own forks in the long-term. There's an increasing maintenance burden we're taking on here that we should avoid if at all possible.

Yea, I get your point and and I agree. 

> I guess it just seems strange that there's a gt_folder parameter which in the original code allows you to specify a parent folder under which you can nest relative paths from gt_ann['file_name'] whereas now the relative paths will be from the working directory.

I can see how that is confusing. This is a trade-off, currently the value of gt_folder is derived from the file_name field in dataset.json. I think the original intention of this code was to have a json file and a predefined folder structure. That way you could have just the file name and provide the gt_folder seperately, based on the predefined folder names. 

If possible, I'd like to refrain from defining things like gt_folders in code as much as possible since it increases the amount of setup required to use different datasets.

That being said, I think there is one other approach that we could try. Given that this code is called through a detectron2 evaluator implementation, we can change the code responsible for registering a dataset to the detectron2 DatasetCatalog (part of our dataset_structures_lib). If we convert relative paths to absolute paths during registration, we should be able to keep relative paths in the dataset.json and we might not need to change code in this forked repo. 

I think it would still differ from the intended use of this code, as it seems to me that the intention is to have only the file name in the file name field. But if it works, I'll take it.

Reply:
Nevermind, on line 190 of `evaluation.py` in this repo, the code opens `gt_json_file`. This is pulled from detectrons MetadataCatalog and corresponds to our `dataset.json`. Even if I change the way the dataset is registered, it won't solve the issue with relative paths.

Looks like our options are to either use absolute paths in the `dataset.json` files, or make changes to a forked repo.